### ISSUE TITLE
Keyboard vibration does not work

### ISSUE THREAD

**Issue by GerryT11** [HelloVolla/ubuntu-touch-beta-tests#77]
System:
Volla Phone with (Ubuntu 16.04 (OTA-14). Last updated 28.09.31 (sic!)

Bug description:
The keyboard vibration (e.g. when typing) does not work. The setting "keyboard vibration" does not have any effect.

Expected behavior:
When the setting "keyboard vibration" is active, the user should "feel" that he/she hits a key on the keyboard.

**Comment by JamiKettunen** [comment]
Are you sure it's actually not vibrating at all when tapping the on-screen keyboard? Do any other vibrations in general work?
The effect works just fine on my unit which is on the latest Development update channel build, **but** it is *very* subtle imho and I'm unsure if it can be made any stronger.

**Comment by GerryT11** [comment]
> Are you sure it's actually not vibrating at all when tapping the on-screen keyboard? Do any other vibrations in general work?
> The effect works just fine on my unit which is on the latest Development update channel build, **but** it is _very_ subtle imho and I'm unsure if it can be made any stronger.

Thanks for your reply. Indeed, I notice some very subtle keyboard vibration. For my thumb, the current vibration is too subtle, I don't feel it. With the index finger, I can feel it, but still very subtle. 
As comparison, Ubuntu Touch on the OnePlus One has stronger keyboard vibration than the Volla Phone. On the OnePlus One it "feels right".

Thus, I changed the title of the bug to: "Make keyboard vibration somewhat stronger"

**Comment by Stevie00911** [comment]
Yes, please make keyboard vibration somewhat stronger.
The vibration effect is almost not noticeable.
To get a good interaction feeling with the device, I find, is important.
With the good old BQ Aquaris E5 with Ubuntu Touch the vibraton feel was ok.

**Comment by mateosalta** [comment]
it should be possible, I think florian had to increase the duration of the opo, so I imagine a similar thing could be done here?

**Comment by Flohack74** [comment]
We basically did smth like this https://github.com/ubports/usensord/pull/3/files to enable an environment variable to give additional vibration time. But, for the Vollaphone this needs to be addressed rather here: https://gitlab.com/ubports/core/hfd-service/-/blob/xenial/src/service.cpp since Halium versions >= 9 should use hfd for vibration effects, not usensord anymore. Basically add a variable that holds the value of the environment variable and then add this value in ms every time a vibrate happens.

**Comment by JamiKettunen** [comment]
Upstream `hfd-service` issue: https://gitlab.com/ubports/development/core/hfd-service/-/issues/6

**Comment by Flohack74** [comment]
I think its an easy fix but not for the current OTA. We can also do it for 16.04 still, should not be too complicated.

**Comment by JamiKettunen** [comment]
There is now a patch ready for testing! From the Terminal app (or `adb shell` on a computer with developer mode active and device connected via USB) run the following:
```bash
sudo su
ubports-qa install PR_hfd-service_10
device=$(getprop ro.product.vendor.device)
mount -o remount,rw /
cat <<EOF > /etc/deviceinfo/devices/$device.yaml
$device:
  VibrateDurationExtraMs: 40
EOF
mount -o remount,ro /
pkill hfd-service
```
Afterwards tapping the on-screen keyboard should restart `hfd-service` automatically and give the vibration some noticeable strength :) This should work on both the Volla Phone (`yggdrasil`) and Volla Phone 22 (`mimameid`), Volla Phone X (`yggdrasilx`) remains untested as I don't have a unit for testing.

Please tweak the `VibrateDurationExtraMs` to your liking an report results here on what the value for which device models should become :)

**Comment by j2g2rp** [comment]
I've tested rightnow on volla phone X with vibrateScale: 40. It works correctly on OSK, it is the same vibration as on waydroid OSK (it wasn't previously)
I've tested the workaround of this bug after this too
https://github.com/HelloVolla/ubuntu-touch-beta-tests/issues/113
Vibration on alarm still doesn't works, so I assume that it is a different bug, isn't it?

Regards

**Comment by JamiKettunen** [comment]
@j2g2rp correct, the alarm vibration not working is a different issue. glad to hear my instructions also worked on VPX and thanks for testing :)

**Comment by Flohack74** [comment]
Can we close this then?

**Comment by JamiKettunen** [comment]
@Flohack74 when https://gitlab.com/ubports/development/core/hfd-service/-/merge_requests/10 is merged I suppose yes :)

As I mentioned it would also need `deviceinfo` configuration however, so actually closing should happen when the configs land in devel channel.

**Comment by j2g2rp** [comment]
just wonder...
It will be included on devel and on next RC?
Just to know if I have test it once back to RC again.

**Comment by JamiKettunen** [comment]
This enhancement will unfortunately be pushed back until OTA 25 at the very earliest, OTA 24 is already packed with stuff and there will be a merge freeze until that is out. You can for the time being however run those commands again to get this fix on any update channel I believe.

**Comment by j2g2rp** [comment]
Ok. No problem. Thanks for the answer!

**Comment by Bolly** [comment]
> There is now a patch ready for testing! From the Terminal app (or `adb shell` on a computer with developer mode active and device connected via USB) run the following:
> 
> ```shell
> sudo su
> ubports-qa install PR_hfd-service_10
> device=$(getprop ro.product.vendor.device)
> mount -o remount,rw /
> cat <<EOF > /etc/deviceinfo/devices/$device.yaml
> $device:
>   vibrateScale: 40
> EOF
> mount -o remount,ro /
> pkill hfd-service
> ```
> 
> Afterwards tapping the on-screen keyboard should restart `hfd-service` automatically and give the vibration some noticeable strength :) This should work on both the Volla Phone (`yggdrasil`) and Volla Phone 22 (`mimameid`), Volla Phone X (`yggdrasilx`) remains untested as I don't have a unit for testing.
> 
> Please tweak the `vibrateScale` to your liking an report results here on what the value for which device models should become :)

hello, after that the VollaPhone22 (Stable channel) does not start. 😥

https://pastebin.com/PBSXdRnS

Now it stays on the "Volla" screen.

_Boot recovery/Factory reset/Format data/ Factory reset/Format data_
```
-- Wiping data...
E: Cannot failed: /metadata/ota: No such file or directory
Formatting /data...
Formatting /metadata...
Data wipe complete.

```

**Comment by JamiKettunen** [comment]
@Bolly ouch, looks like `ubports-qa` in general should not be used on stable update channel then :/ I think you already found out that [the UBports installer](https://devices.ubuntu-touch.io/installer/) has support to reflash Ubuntu Touch on the VP22 if needed.

**Comment by JamiKettunen** [comment]
This was tracked on https://ubports.atlassian.net/browse/YUMI-12 and should now be available on both latest 16.04 development channel update and 20.04 on all models :)
